### ISSUE TITLE
Environment.ProcessPath / Environment.GetCommandLineArgs()[0] discrepancies

### ISSUE THREAD

**Issue by xoofx** [dotnet/runtime#101837]
Hello there, ☺️ 

This issue is related to the following previous issues:

* #11305
* #30212

`Environment.ProcessPath` and `Environment.GetCommandLineArgs()[0]` are behaving quite differently (See the follow parts for more details):
- between them
- between platforms
- between regular dotnet built app and NativeAOT
- When we access the exe through a symbolic link

This is quite problematic when trying to build apps that expect to recover the C/C++ equivalent of `argv[0]` which should give exactly what was typed on the command line (e.g if it is a symbolic link, we should keep its name), and not something else.

# Reproducible steps

## Windows - .NET 8.0

Powershell with admin rights (for the symbolic link):

```pwsh
dotnet new console -n HelloWorld
cd HelloWorld
Write-Output "Console.WriteLine($""Environment.ProcessPath: {Environment.ProcessPath}"");" > Program.cs
Write-Output "Console.WriteLine($""Environment.GetCommandLineArgs()[0]: {Environment.GetCommandLineArgs()[0]}"");" >> Program.cs
dotnet build -c Release
cd .\bin\Release\net8.0
```

The we create a symbolic link from `special_hello.exe` to `HelloWorld.exe`:

```pwsh
New-Item -ItemType SymbolicLink -Path special_hello.exe -Value HelloWorld.exe
.\special_hello.exe
```

Outputs:

```
Environment.ProcessPath: C:\code\HelloWorld\bin\Release\net8.0\special_hello.exe
Environment.GetCommandLineArgs()[0]: C:\code\HelloWorld\bin\Release\net8.0\HelloWorld.dll
```

## Windows - NativeAOT .NET 8.0

Similar steps then before, but with `dotnet publish -c Release -p:PublishAot=true`

```shell
cd .\bin\Release\net8.0\win-x64\publish
New-Item -ItemType SymbolicLink -Path special_hello.exe -Value HelloWorld.exe
.\special_hello.exe
```

Outputs:

```
Environment.ProcessPath: C:\code\HelloWorld\bin\Release\net8.0\win-x64\publish\special_hello.exe
Environment.GetCommandLineArgs()[0]: C:\code\HelloWorld\bin\Release\net8.0\win-x64\publish\special_hello.exe
```

## MacOS/Ubuntu - .NET 8.0

```shell
cd ./bin/Release/net8.0
ln -s HelloWorld special_hello
./special_hello
```

Outputs:

```
Environment.ProcessPath: /home/xoofx/code/HelloWorld/bin/Release/net8.0/HelloWorld
Environment.GetCommandLineArgs()[0]: /home/xoofx/code/HelloWorld/bin/Release/net8.0/HelloWorld.dll
```

## MacOS/Ubuntu - NativeAOT .NET 8.0

```shell
cd ./bin/Release/net8.0/linux-x64/publish
ln -s HelloWorld special_hello
./special_hello
```

Outputs:

```
Environment.ProcessPath: /home/xoofx/code/HelloWorld/bin/Release/net8.0/linux-x64/publish/HelloWorld
Environment.GetCommandLineArgs()[0]: ./special_hello
```

# Expected behavior

## `Environment.GetCommandLineArgs()[0]`

This value should always be the strict equivalent of `argv[0]` in a C/C++ program (and `MacOS/Ubuntu - NativeAOT .NET 8.0` is the only one to respect it from the above)

In particular:
- It should not transform to an absolute path or resolve symbolic link
- It should not give a dll whether we are running through corehost from a regular `dotnet build`

## `Environment.ProcessPath`

From the above it is unclear if:

- It should give an absolute path to the exe name but not following symbolic link? (as on `Windows - NativeAOT .NET 8.0`) e.g `C:\code\HelloWorld\bin\Release\net8.0\win-x64\publish\special_hello.exe`
- It should give an absolute path to the fully resolve exe name following symbolic link? (as on MacOS/Ubuntu) e.g `/home/xoofx/code/HelloWorld/bin/Release/net8.0/HelloWorld`

# Implementation problems

I tried to fix it but the problem is that we are not able to recover `argv[0]` from `coreclr_execute_assembly`, and I end up hacking `argv` to actually have `argv[0]` in it and putting 1 on bit 0 on the pointer to understand behind that it is a special `argv`, which I conceded is super ugly 😅

```
int coreclr_execute_assembly(
            void* hostHandle,
            unsigned int domainId,
            int argc,
            const char** argv,
            const char* managedAssemblyPath,
            unsigned int* exitCode)
```

and while `coreclr_initialize` receive an `exePath` (which is not `argv[0]`), it is processing it to create a commandline string and things get more convoluted here (and it's not used by `Environment.GetCommandLineArgs()[0]`)

```
int coreclr_initialize(
            const char* exePath,
            const char* appDomainFriendlyName,
            int propertyCount,
            const char** propertyKeys,
            const char** propertyValues,
            void** hostHandle,
            unsigned int* domainId)
```

I assume that it is not possible to change the signature of these functions?

I would be happy to make a PR, but some guidance regarding `coreclr_execute_assembly`/`coreclr_initialize` on how to propagate `argv[0]` would be appreciated. (If I can add a new arg, that should be relatively easy to fix - but I would have change also e.g `CorHost2::ExecuteAssembly`)

Thanks! ☺️ 

cc: @jkotas @MichalStrehovsky @agocke

**Comment by jkotas** [comment]
- Consider what the behavior should be for invocation using `dotnet` host, e.g. `dotnet foo.dll`. The current behavior of `GetCommandLineArgs` tries to be consistent between `dotnet hello.dll` and `hello.exe`.

- Changing the behavior of `GetCommandLineArgs()[0]` would need to be treated as a breaking change. For example, code our there can be doing things like `AssemblyName.GetAssemblyName(Environment.GetCommandLineArgs()[0])`.

- `Environment.ProcessPath`: We resolve the symlinks on some platforms because we have to in order to return absolute path.

**Comment by xoofx** [comment]
> Consider what the behavior should be for invocation using dotnet host, e.g. dotnet foo.dll. The current behavior of GetCommandLineArgs tries to be consistent between dotnet hello.dll and hello.exe.

Indeed, something I should probably add in the differences.

> Changing the behavior of GetCommandLineArgs()[0] would need to be treated as a breaking change. For example, code our there can be doing things like AssemblyName.GetAssemblyName(Environment.GetCommandLineArgs()[0]).

From the above, it looks like `Environment.GetCommandLineArgs()[0]` is so unspecified that it is barely possible to use it in a consistent manner across NativeAOT/Regular App/Platforms. So I could definitely see some folks relying on it for specific niche cases, but I see also plenty of folks not able to use it because of these inconsistencies (I would personally not use it at all)

If this is not desirable to make a breaking change, would it be not better then to add a new property to `Environment`, like `Environment.Argv0` or something similar? I would make a proposal then. Ideally, I would prefer to introduce a different property for the case of e.g `dotnet foo.dll` because it would then make `Environment.GetCommandLineArgs()` the strict equivalent of `argv`, which is the confusing part today. But if this breaking change is not possible, then we should consider a new property.

> Environment.ProcessPath: We resolve the symlinks on some platforms because we have to in order to return absolute path.

The doc says `Returns the path of the executable that started the currently executing process`. Is there a technical requirement that it should point to the resolved exe?

**Comment by jkotas** [comment]
> it looks like Environment.GetCommandLineArgs()[0] is so unspecified that it is barely possible to use it in a consistent manner across NativeAOT/Regular App/Platforms.

Yes, `GetCommandLineArgs()[0]` is problematic for single file deployments (that includes NativeAOT). Single file deployments are a niche. Outside of single file deployments, `GetCommandLineArgs()[0]` always returns the managed binary that started the process.

> Is there a technical requirement that it should point to the resolved exe?

It should be an absolute path and it should not be a placeholder like `/proc/self/exe`.

**Comment by xoofx** [comment]
Looking at the usage of [`Environment.GetCommandLineArgs()[0]`](https://github.com/search?q=%22Environment.GetCommandLineArgs%28%29%5B0%5D%22&type=code), it is definitely used a lot, so changing it would break too many things. (Not all the usages are expecting a DLL, some are expecting an exe name, but that depends on their usage for sure).

Maybe introducing a new API that would give the additional parts as prefixes would help:

`string? Environment.PreCommandLineArg0` would return:

| Exe Name | `Environment.PreCommandLineArg0` | `Environment.GetCommandLineArgs()` 
|------------------|--------------|---------------
| `./dotnet.exe foo.dll 1 2 3`| `./dotnet.exe`  | `foo.dll`, `1`, `2`, `3` |
| `./foo.exe 1 2 3`| `./foo.exe`  | `foo.dll`, `1`, `2`, `3`
| Regular .NET App `./special_foo.exe 1 2 3` special_foo.exe symlink | `./special_foo.exe` | `foo.dll`, `1`, `2`, `3`
| NativeAOT .NET App `./special_foo.exe 1 2 3` special_foo.exe symlink | null | `./special_foo.exe`, `1`, `2`, `3`

with `Environment.PreCommandLineArg0` being the true `argv[0]` equivalent of C/C++ when it is not covered by `Environment.GetCommandLineArgs()` for NativeAOT case.

Would that be an acceptable API to resolve this situation?

**Comment by xoofx** [comment]
Bringing some folks from other issues that would be interested in solving this: @rjmholt, @tmds, @mklement0

**Comment by jkotas** [comment]
I am not following why we would want the API to return null for native AOT.

If we were to add a new API for this, I think it should be a policy-free API that does not change its behavior based on how the process was invoked. `string? Environment.ProcessInvocationName`. Returns the name used to invoke the process, same as `argv[0]` in C/C++, in all situations. Returns null when the invocation name is not available.

**Comment by xoofx** [comment]
I have created the following API proposal for this issue:

- #101872

**Comment by tmds** [comment]
Note that there is also: `Process.GetCurrentProcess().MainModule.FileName`.

> Environment.ProcessPath

Perhaps we could initialize this to the full path for `argv[0]` without resolving links?

It wouldn't allow to get the exact `argv[0]` but it might enable all desired use-cases without adding another API.

**Comment by xoofx** [comment]
> It wouldn't allow to get the exact argv[0] but it might enable all desired use-cases without adding another API.

Not all use-cases, like login shell as you highlighted [here](https://github.com/dotnet/runtime/issues/30212#issuecomment-512330585) with the character `-` added as a prefix when running a shell as a login shell. The purpose here is to get the real `argv[0]`.

Also, we are still missing the case of `dotnet foo.dll` where `foo.dll` is put in `Environment.GetCommandLineArgs()[0]` which could be detected with `Environment.IsImplicitCommandLineArg0` returning false.

**Comment by tmds** [comment]
> Not all use-cases, like login shell as you highlighted https://github.com/dotnet/runtime/issues/30212#issuecomment-512330585 with the character - added as a prefix when running a shell as a login shell. The purpose here is to get the real argv[0].

We may want to accept the behavior for this very specific case.

An other option may be: change `GetCommandLineArgs()[0] ` so it contains `argv[0]` when it is executed through an app host and remains the current behavior when executed through the `dotnet` host (full path to managed assembly).

**Comment by xoofx** [comment]
> An other option may be: change GetCommandLineArgs()[0]  so it contains argv[0] when it is executed through an app host and remains the current behavior when executed through the dotnet host (full path to managed assembly).

I would have preferred as well but that would break too many expected behaviors as based on usage on GitHub [here](https://github.com/dotnet/runtime/issues/101837#issuecomment-2093218661)

**Comment by tmds** [comment]
> I would have preferred as well but that would break too many expected behaviors as based on usage on GitHub here

We would break some of these and also fix some others.

If you're now using `GetCommandLineArgs()[0]` you have a bug depending on how the app was published.

**Comment by jkotas** [comment]
> We would break some of these and also fix some others.

Are there any examples of code out there that would be fixed by changing `GetCommandLineArgs()[0]` behavior to match `argv[0]` unless the app was launched via a host?

`GetCommandLineArgs()[0]` in .NET Framework matches `argv[0]`:
```
C:\repro>type test.cs
using System;

class My {
    static void Main() {
        Console.WriteLine(Environment.GetCommandLineArgs()[0]);
    }
}
C:\repro>test.exe
test.exe

C:\repro>c:\repro\test.exe
c:\repro\test.exe
```

So we can change would be undoing an (unintentional?) .NET Framework -> .NET Core change. Looking at the github examples, I think the breaking change should be acceptable.

**Comment by jkotas** [comment]
| Command line | `Environment.GetCommandLineArgs()` 
|------------------|---------------
| `dotnet.exe foo.dll 1 2 3`| `foo.dll`, `1`, `2`, `3` |
| `foo 1 2 3`| `foo`, `1`, `2`, `3`
| `../bar/special_foo.exe 1 2 3` with `special_foo.exe` symlinked to `foo.exe`| `../bar/special_foo.exe`, `1`, `2`, `3`

**Comment by tmds** [comment]
> So we can treat it as undoing an (unintentional?) .NET Framework -> .NET Core change. Looking at the github examples, I think the breaking change should be acceptable.

Yes, many of the samples expect the .NET Framework semantics: something a user can execute, and a path that may be relative.

**Comment by xoofx** [comment]
> Are there any examples of code out there that would be fixed by changing GetCommandLineArgs()[0] behavior to match argv[0] unless the app was launched via a host?

Difficult to say because `GetCommandLineArgs()[0]` is unusable today for the argv[0] scenario. I started this issue because I was just developing a command line parser library for C#, and I have left the code using `Environment.ProcessPath` instead of `GetCommandLineArgs()[0]` for this reason:

https://github.com/XenoAtom/XenoAtom.CommandLine/blob/c3e2d29389a39ea8d5335d6ec265f25eedec95c7/src/XenoAtom.CommandLine/CommandApp.cs#L94-L98

> So we can treat it as undoing an (unintentional?) .NET Framework -> .NET Core change. Looking at the github examples, I think the breaking change should be acceptable.

If we could do that, I would be happy to make a PR. I believe that the case of `dotnet.exe foo.dll` can be detected by checking that the `GetCommandLineArgs()[0]` is actually a dll (and additionally that `Environment.ProcessPath` is the `dotnet.exe`).

**Comment by xoofx** [comment]
Pushed a prototype fix #101880, not really thrilled by it, but I had some difficulty to pass this information over without changing any of the existing CorHost2 APIs.

**Comment by tmds** [comment]
It would be nice if we can fix this for .NET 9.

**Comment by xoofx** [comment]
> It would be nice if we can fix this for .NET 9.

Agreed, I will try to come back with a better PR in the coming months.

**Comment by am11** [comment]
Another potential solution could be to introduce a new API, `string[] Environment.NativeCommandLineArgs`, which would expose the raw command line arguments as received from the hosting layer (corehost,nativeaot etc.). These arguments could then be accessed via a QCall. While this approach may not fully resolve all discrepancies, it could address use cases which require the unmodified arguments (with the exception of differences in `argv` processing rules between POSIX and Win32 systems).

**Comment by xoofx** [comment]
Hello folks, I gave another try via #131671 to solve this issue.
